modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -37,12 +42,18 @@
 
     private final Netty4Transport transport;
 
+    private final ConcurrentLinkedQueue<WriteOperation> queuedWrites = new ConcurrentLinkedQueue<>();
+
+    private volatile ChannelHandlerContext ctx;
+
+    private WriteOperation currentWrite;
+
     Netty4MessageChannelHandler(Netty4Transport transport) {
```

Comment:
I'm not sure that this should be in this (`Netty4MessageChannelHandler`). You are copying heap bytes -> direct bytes. That works fine for plaintext. But this handler comes before the `SslHandler`. Which means in the security use case, this is heap bytes -> direct bytes -> heap bytes (temporary byte array in `SSLEngine`) encrypted->  heap buffer -> direct bytes (in `filterOutboundMessages`). So this will introduce more copies.

I think that either:

1. This should be a outbound handler that comes after the `SslHandler`
2. This should not copy the bytes to a direct buffer and instead just limit the heap bytes that are passed to the next handler until the writability callback is invoked.

Reply:
Right :) I went with option 2. now. That also makes the whole issue of [potentially trying to write released bytes](https://github.com/elastic/elasticsearch/pull/39286#discussion_r261422662) a lot safer/cleaner. Once/if the buffer is below our write chunk size we can simply pass it down outright and use retained slices in the writes preceding that situation.
(+ we save some objects for writes that are smaller than the buffer size :))